### ISSUE TITLE
Associate inferred stencil specifications to a complete data reference rather than just the label

### ISSUE THREAD

**Issue by dorchard** [camfort/camfort#59]
See for example:
```
  ! subscript on data ref
  do i=1,10
     c%yval(i) = d%yval(i)
  end do
```
(from https://github.com/camfort/camfort/blob/master/samples/stencils/dataRefExample2.f90)
We infer specification:
```
(33:6)-(33:26)    stencil readOnce, (pointed(dim=1)) :: yval
```
But it should be associated to `d%yval`.

**Comment by GuiltyDolphin** [comment]
Closing as FORTRAN standards indicate that the parsing is currently correct (`d%yval(i)` -> `d % (yval(i))`. See section 6.1.2 of ftp://ftp.nag.co.uk/sc22wg5/N001-N1100/N692.pdf

**Comment by dorchard** [comment]
While the parsing seems correct - I still think this is the wrong kind of label to associate for the stencil specification.

**Comment by dorchard** [comment]
Although this is very minor so it's probably not a big issue.

**Comment by GuiltyDolphin** [comment]
@dorchard I'm a bit confused, given the parsing is already correct, what sort of stencil would you expect to see? This `(33:6)-(33:26)    stencil readOnce, (pointed(dim=1)) :: d%yval`? Or something different?

**Comment by madgen** [comment]
Yeah, that's what he expects. `yval` is only a field name. Consider the following program

```fortran
program hey
  type type1
    integer, dimension(10) :: array
  end type type1

  type type2
    integer, dimension(10) :: array
  end type type2

  type (type1) x
  type (type2) y

  integer i

  do i = 1, 10
    x % array(i) = x % array(i) + y % array(i-1)
  end do
end program hey
```

If you run stencil infer, you get

```
(16:5)-(16:33)    stencil readOnce, (backward(depth=1, dim=1, nonpointed)) :: array
```

If you remove the term ` y % array(i-1)`, then you get

```
(16:5)-(16:31)    stencil readOnce, (pointed(dim=1)) :: array
```

So not only it doesn't make sense to qualify field name with a stencil, but it also overwrites a valid spec instead of producing an additional one.

**Comment by GuiltyDolphin** [comment]
@dorchard @madgen Looking at https://courses.physics.illinois.edu/phys466/sp2013/comp_info/derived.html the example seems to indicate that you are correct. I think I misunderstood what the grammar (and constraints) were saying in the linked document - Harry and I have gone over it and I'm now pretty sure that our parsing is wrong (@dorchard as discussed).

**Comment by madgen** [comment]
@GuiltyDolphin  Why is our parsing wrong?

Regardless the parsing is wrong or not this bug is not related to parsing. It is related to grabbing wrong AST node (too deep) to associate a stencil to. It should be the outermost `ExpDataRef`. Another resolution is to change the renamer such that each field gets a fully qualified name according to the left of `%`, but that is more of a hack than anything else.

**Comment by GuiltyDolphin** [comment]
@madgen I believe we are parsing the expression as `d%(yval(i))`, so we are treating `yval` as an array rather than a key into `d` (@dorchard correct me if I'm wrong - was that what we were seeing from `camfort ast`?). 

The tests have a case for `x % y(10) % z` and seem to be checking it parses as `(x % (y(10))) % z`, which I think must be wrong if we are accessing `(10)` at `y` in `x` (I believe it should be `((x % y)(10)) % z` instead.

I'll have more of a look tomorrow, and find out what we're grabbing!

**Comment by madgen** [comment]
Hmm, yes. I see your point. You might find it difficult to unambiguously parse such expressions, but fingers crossed. Though that should be made an issue in fortran-src not in CamFort. The issue in CamFort as I explained previously is orthogonal to this one.

Also you can use `$ fortran-src -a parse <filename>` to see an AST dump.
